EIPS/eip-2121.md

```diff
@@ -0,0 +1,81 @@
+* Gas cost of calling/using registry style contracts including token contracts can be greatly reduced
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+1. Currently when off-chain applications want to consume data from smart contracts, they have to use the existing ABI function calls defined in the corresponding contracts. This means that these off-chain consumers must have a) a full & up-to-date EVM and b) download the contracts storage data locally to execute and fetch the result of the ABI calls -- or that these consumers **must fallback to use a centralized gateway** that can execute an ABI call for them such as with the "eth_call" RPC API. A work-around to this has been to **hard-code in off-chain clients the actual memory layout** of the compiled smart contract storage and to access variables directly, validating the storage against the merkle tree. This works with careful engineering but is very brittle since the actual memory slot position of smart contract variables can change easily when there are compiler changes, addition/removal of smart contract variables or variable re-ordering. Also this mechanism can not be expressed in a common interface such an ERC as slot calculations need to be made on a per contract basis.
+
+2. The gas cost when interacting with registry style contract such as ERC-20s, ENS or a membership lists is very high, because they incur cross contract call cost, cost for ABI decoding and dispatching and finally loading the data. In many cases the underlying storage that is being queried is though just a simple mapping. In these cases a new SLOAD2 call directly accessing the mapping in storage could not only **reduce the gas cost** of the interaction more than 10x, but also it would make the gas cost **predictable** for the reading contract.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+**Proposal**
+
+There are two additions to solidity and one new opcode in the EVM:
+
+* A new EVM instruction `SLOAD2 (0x5c)` that works like `SLOAD (0x54)` with the same gas cost but has an additional parameter representing the contract that is to be read from.
+
+```
+SLOAD (0x5c)
+```
+
+* A "fixed(@N)" solidity keyword that marks a contract storage variable as externally readable. The @N is a storage slot address and forces the compiler to put the contract variable at the specified slot N. The data position is thus becoming part of the interface.
```

Comment:
I think Solidity syntax should be discussed on https://github.com/ethereum/solidity and EIPs changing the EVM have little say in the syntax. They can suggest things, but not as part of a specification.

Reply:
Additionally there is a very lengthy discussion on the topic of selecting storage slots in the language: https://github.com/ethereum/solidity/issues/597

Reply:
@axic is there something like a solidity EIP process or should I just open an issue?

Reply:
There is no such thing as a Solidity EIP process. Design discussion about the language are not handling in this repo. I'd suggest to open an issue or continue discussion under https://github.com/ethereum/solidity/issues/597.

It is perfectly fine giving examples or explaining what Solidity should do as part of this EIP, but a Solidity specification doesn't belong here. Additionally restricting this EIP to EVM changes only makes the scope smaller, easier to understand and reason about.

Reply:
ok, will split and open a solidity issue